Framework/API/src/MatrixWorkspace.cpp

```diff
@@ -180,35 +180,76 @@ const std::string MatrixWorkspace::getTitle() const {
+/** Return a reference to the SpectrumInfo object.
  *
- * By default ThreadedContextCheck::Check enables a check of the OpenMP thread
- * number to actively fail when called from a threaded context. In cases where
- * the code is calling this in a threaded context but with different workspaces
- * this check can be disabled by setting contextCheck =
- * ThreadedContextCheck::Skip. This is not recommended and should be used with
- * extreme care. This flag is a temporary workaround and will be removed once
- * Instrument-2.0 is introduced.
- *
- * Currently this method is not thread-safe, and any modifications of the
- * instrument or instrument parameters such as mask flags will not be visible in
- * existing SpectrumInfo references. After modification, obtain a new reference
- * by calling this method again.
+ * Any modifications of the instrument or instrument parameters will invalidate
+ * this reference.
+ */
+const SpectrumInfo &MatrixWorkspace::spectrumInfo() const {
+  if (!m_spectrumInfo) {
+    std::lock_guard<std::mutex> lock{m_spectrumInfoMutex};
+    if (!m_spectrumInfo)
+      m_spectrumInfo = Kernel::make_unique<SpectrumInfo>(*this);
+  }
+  return *m_spectrumInfo;
+}
+
+/** Return a non-const reference to the SpectrumInfo object. Not thread safe.
  */
-const SpectrumInfo &
-MatrixWorkspace::spectrumInfo(ThreadedContextCheck contextCheck) const {
-  if (contextCheck == ThreadedContextCheck::Check &&
-      PARALLEL_THREAD_NUMBER != 0)
-    throw std::runtime_error("MatrixWorkspace::spectrumInfo(): Thread ID is "
-                             "not 0. This indicates that the method is called "
-                             "in a threaded context. This is not allowed.");
-
-  // For now we *always* create a new SpectrumInfo since the instrument or
-  // parameters may have changed.
+SpectrumInfo &MatrixWorkspace::mutableSpectrumInfo() {
+  // SpectrumInfo holds a parameterized instrument so the reference count to the
+  // ParameterMap in ExperimentInfo is never 1, i.e., calling
+  // ExperimentInfo::instrumentParameters (as needed for many modifications)
+  // would always trigger a copy and leave the instrument information in
+  // SpectrumInfo in a detached state. Here we work around this as follows:
+  //
+  // By creating a new SpectrumInfo any previous references that we are
+  // responsible for are released. Creating SpectrumInfo with a non-const
+  // reference to a MatrixWorkspace will cache to result of
+  // ExperimentInfo::instrumentParameters() which will later be used by
+  // modifications. This will trigger a copy if required. SpectrumInfo will
+  // obtain the parameterized instrument after obtaining the ParameterMap such
+  // that both refer to the same instance. This is the same method existing
+  // algorithms use to avoid an extra copy (call
+  // MatrixWorkspace::instrumentParameters() then
+  // MatrixWorkspace::getInstrument()).
+  m_spectrumInfo = nullptr;
```

Comment:
No:

`SpectrumInfo` caches a parameterized copy of the instrument (obtained via `MatrixWorkspace::getInstrument()`). The reference count to the `ParameterMap` is thus at least 2 (1 from the workspace, 1 from `SpectrumInfo`).

Below we want to create a `SpectrumInfo` from a non-const workspace. That will get the non-const parameter map from the workspace. If the reference count is larger than 1 this will make a copy of the parameter map (copy-on-write).

To avoid unnecessary copies we are thus first releasing the old `SpectrumInfo`, dropping the reference count to the parameter map by 1. If the workspace is the sole remaining owner `make_unique<SpectrumInfo>(*this)` will then NOT trigger a copy.